File: 2025/kotlin/src/main/kotlin/de/rubixdev/aoc/Day01.kt
package de.rubixdev.aoc

import kotlin.math.absoluteValue
import kotlin.math.sign

fun day1(input: String): Day = sequence {
    val input = input.lines()
        .map { line -> line.replace("R", "").replace('L', '-').toInt() }

    yield(null)
    yield(part1(input))
    yield(part2(input))
}

private fun part1(input: List<Int>): Int = input
    .asSequence()
    .scan(50) { acc, i -> (acc + i).mod(100) }
    .count { it == 0 }

private fun part2(input: List<Int>): Int = input
    .asSequence()
    .flatMap { n -> sequence { repeat(n.absoluteValue) { yield(n.sign) } } }
    .scan(50) { acc, i -> (acc + i).mod(100) }
    .count { it == 0 }

File: 2025/scala/src/main/scala/de/rubixdev/aoc/day01.scala
package de.rubixdev.aoc

def day1(rawInput: String): Unit = {
  val input = rawInput.linesIterator
    .map(_.replace("R", "").replace('L', '-').toInt)
    .toList

  println("Part 1: " + part1(input))
  println("Part 2: " + part2(input))
}

private def part1(input: List[Int]): Int = input.view
  .scan(50) { (acc, i) => (acc + i) % 100 }
  .count(_ == 0)

private def part2(input: List[Int]): Int = input.view
  .flatMap { n => LazyList.fill(n.abs) { n.sign } }
  .scan(50) { (acc, i) => (acc + i) % 100 }
  .count(_ == 0)
